src/mrack/providers/aws.py

```diff
@@ -209,9 +212,87 @@ async def validate_hosts(self, reqs):
+                f"{log_msg_start} Error retrieving info from subnet: {subnet_id}"
+            )
+            return 0
+
+        logger.debug(
+            f"{log_msg_start} Subnet Id: {subnet_id}\n"
+            f"\tHas {subnet.available_ip_address_count} IP addresses available"
+        )
+
+        return subnet.available_ip_address_count
+
+    async def can_provision(self, hosts):  # pylint: disable=too-many-branches
+        """Check that all host can be provisioned.
+
+        Checks:
+        * Available IPv4 addresses are enough
+        """
+        log_msg_start = self.dsp_name
+
+        ip_availabilities = {}  # Dict for storing available IPs in every subnet
+        single_subnet_prov = []  # List for storing provided single subnets
+        mult_subnets_prov = []  # List for storing lists of provided multiple subnets
+
+        # For all hosts, get subnets in dict (with no repetitions) and list of
+        # provided single & multiple subnets
+        for host in hosts:
+            if subnet_ids := host.get("subnet_ids"):
+                if len(subnet_ids) == 1:
+                    single_subnet_prov.append(subnet_ids[0])
+                    if subnet_ids[0] not in ip_availabilities:
+                        ip_availabilities[subnet_ids[0]] = 0
+                else:
+                    mult_subnets_prov.append(subnet_ids)
+                    for subnet_id in subnet_ids:
+                        if subnet_id not in ip_availabilities:
+                            ip_availabilities[subnet_id] = 0
+            else:
+                logger.debug(
+                    f"{log_msg_start} No subnet/s specified for host {host['name']}."
+                )
+
+        # Get available IPs from AWS for every subnet
+        ips_count_wait = [
+            self.get_subnet_available_ips(subnet_id, log_msg_start)
+            for subnet_id in ip_availabilities
+        ]
+        ips_count_result = await asyncio.gather(*ips_count_wait)
+        if not all(ips_count_result) > 0:
+            return False
+
+        ip_availabilities = dict(zip(ip_availabilities, ips_count_result))
+        self.subnets_capacity = ip_availabilities.copy()
+
+        # Discount hosts with single subnet specified from available IPs
+        for subnet_id in single_subnet_prov:
+            ip_availabilities[subnet_id] -= 1
+
+        # Discount hosts with multiple subnets specified from available IPs
+        no_ips_flag = True
+        for subnet_ids in mult_subnets_prov:
+            for subnet_id in subnet_ids:
+                if ip_availabilities[subnet_id] > 0:
+                    ip_availabilities[subnet_id] -= 1
+                    no_ips_flag = False
+                    break
+            if no_ips_flag:
+                logger.info(
+                    f"{log_msg_start} Not enougn IP addresses available "
+                    f"in subnet/s: {subnet_ids}"
+                )
+                return False
+            no_ips_flag = True
```

Comment:
With this implementation, it's supposed to check every subnet provided (in subnet_ids) before deciding that it can't be provisioned, that's why I wanted to break to the outer loop.
I have refactored the code using `For-Else` [syntax](https://book.pythontips.com/en/latest/for_-_else.html#for-else) to get rid of the variable.

Reply:
Thanks